backend/src/test/java/service/UserServiceTest.java

```diff
@@ -0,0 +1,535 @@
+
+        ArgumentCaptor<IncorrectLoginsDto> argument = ArgumentCaptor.forClass(IncorrectLoginsDto.class);
+        verify(incorrectLoginsService).update(argument.capture());
+        assertEquals(result, expected);
+        assertEquals(1, argument.getValue().getIncorrectAttempts());
+    }
+
+    @Test
+    /*
+     * Preconditions: Username is invalid and is the first invalid attempt
+     * Post-conditions: Returns JSONObject indicating invalid login attempt and increments incorrect attempt count
+     */
+    public void testInvalidLoginFirstTime() {
+        User user = mockUser();
+        UserDto userDto = mockUserDto();
+        JSONObject expected = new JSONObject();
+
+        expected.put(BeapEngineConstants.SUCCESS_STR, false);
+        expected.put("status_code", HttpStatus.UNAUTHORIZED.value());
+        expected.put("message", "invalid username or password");
+
+        when(userDao.findByUserPass(user.getUsername(), user.getPassword())).thenReturn(null);
+        when(userDao.findByUsername(user.getUsername())).thenReturn(user);
+        when(userMapper.model2Dto(Mockito.<User>anyObject(), Mockito.<UserDto>anyObject())).thenReturn(userDto);
+
+        when(incorrectLoginsService.getByUserId(Mockito.anyLong())).thenReturn(null);
+
+        JSONObject result = userService.login(user.getUsername(), user.getPassword());
+
+        ArgumentCaptor<IncorrectLoginsDto> argument = ArgumentCaptor.forClass(IncorrectLoginsDto.class);
+        verify(incorrectLoginsService).save(argument.capture());
+        assertEquals(result, expected);
+        assertEquals(1, argument.getValue().getIncorrectAttempts());
+        assertEquals(userDto, argument.getValue().getUserDto());
+        assertFalse(argument.getValue().getLocked());
+        assertNull(argument.getValue().getLockedDate());
+    }
+
+    @Test
+    /*
+     * Preconditions: Invalid login and user account is already locked and unlock time has not past
+     * Post-conditions: Returns JSONObject indicating account is locked
+     */
+    public void testInvalidLoginLocked() {
+        User user = mockUser();
+        UserDto userDto = mockUserDto();
+        JSONObject expected = new JSONObject();
+
+        expected.put(BeapEngineConstants.SUCCESS_STR, false);
+        expected.put("status_code", HttpStatus.LOCKED.value());
+        expected.put("message", "Your account is locked for 24 hours.");
+
+        when(userDao.findByUserPass(user.getUsername(), user.getPassword())).thenReturn(null);
+        when(userDao.findByUsername(user.getUsername())).thenReturn(user);
+        when(userMapper.model2Dto(Mockito.<User>anyObject(), Mockito.<UserDto>anyObject())).thenReturn(userDto);
+        when(util.dateDifference(Mockito.<Date>any(), Mockito.<Date>any())).thenReturn(2L);
+        when(incorrectLoginsService.getByUserId(Mockito.anyLong())).thenReturn(
+                mockIncorrectLoginsDto(
+                        userDto,
+                        true,
+                        0,
```

Comment:
would be larger than 0 but is not actually checked in this logic.

Reply:
ill change it for better understanding of our testing